src/code42cli/cmds/auditlogs.py

```diff
@@ -153,43 +200,62 @@ def send_to(
         affected_user_ids=affected_user_id,
         affected_usernames=affected_username,
     )
+    if use_checkpoint:
+        events = _dedupe_checkpointed_events_and_store_updated_checkpoint(
+            cursor, use_checkpoint, events
+        )
+    with warn_interrupt():
+        event = None
+        for event in events:
+            logger.info(event)
+        if event is None:  # generator was empty
+            click.echo("No results found.")
 
 
-def _search(sdk, format, **filter_args):
-
-    formatter = OutputFormatter(format, AUDIT_LOGS_DEFAULT_HEADER)
+def _get_all_audit_log_events(sdk, **filter_args):
     response_gen = sdk.auditlogs.get_all(**filter_args)
-
     events = []
     try:
-        for response in response_gen:
-            response_dict = json.loads(response.text)
-            if EVENT_KEY in response_dict:
-                events.extend(response_dict.get(EVENT_KEY))
+        responses = list(response_gen)
     except KeyError:
         # API endpoint (get_page) returns a response without events key when no records are found
         # e.g {"paginationRangeStartIndex": 10000, "paginationRangeEndIndex": 10000, "totalResultCount": 1593}
-        pass
+        # we can remove this check once PL-93211 is resolved and deployed.
+        return events
 
-    event_count = len(events)
-    if not event_count:
-        click.echo("No results found.")
-    elif event_count > 10:
-        click.echo_via_pager(formatter.get_formatted_output(events))
-    else:
-        formatter.echo_formatted_list(events)
+    for response in responses:
+        if EVENT_KEY in response.data:
+            response_events = response.data.get(EVENT_KEY)
+            events.extend(response_events)
 
+    return sorted(events, key=lambda x: x.get("timestamp"))
 
-def _send_to(sdk, hostname, protocol, format, **filter_args):
-    logger = get_logger_for_server(hostname, protocol, format)
-    with warn_interrupt():
-        response_gen = sdk.auditlogs.get_all(**filter_args)
-        try:
-            for response in response_gen:
-                if EVENT_KEY in response:
-                    for event in response[EVENT_KEY]:
-                        logger.info(event)
-                else:
-                    logger.info("No results found.")
-        except KeyError:
-            pass
+
+def _dedupe_checkpointed_events_and_store_updated_checkpoint(
+    cursor, checkpoint_name, events
+):
+    checkpoint_events = cursor.get_events(checkpoint_name)
+    new_timestamp = None
+    new_events = []
+    for event in events:
+        event_hash = hash_event(event)
```

Comment:
since the checkpoint that gets stored finally is the last event, why not just set the timestamp of the last event rather than do it one by one for all events. Calling `list` over this function will cause the generator to execute completely.

Reply:
I made `_dedupe_checkpointed_events_and_store_updated_checkpoint` a generator because `send-to` sends events one at a time to the server, and in the event of a network disconnect or other exception during sending, the checkpoint will remain accurate up to the last successful event processed, since we'd error out mid-generator. 

For `search` where we're just echoing to stdout (and need to know the length of the event list to determine _how_ we're echoing), we have to call list() on it first.

And it is possible that the last two events have the same timestamp, which if we want to accurately deduplicate across runs when events have the same timestamp, we need to store both of those last event hashes in the cursor, so we need to process every event in the list to be sure.

Reply:
I'll add a docstring to the function to make that intention clear for future readers.